### PR TITLE
Documentation/Config changes (proposed) for #6589

### PR DESCRIPTION
**Pull request by twghu**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message:
Documentation and Config changes as per request https://github.com/envoyproxy/envoy/issues/6589#issuecomment-622120041 

NOTE: Incomplete, **do not merge** this PR was incorrectly attributed to v2 and v2 is frozen.

Additional Description:
These changes are not the complete fix.
Risk Level: 
Testing:
Docs Changes:
_rst_ and _proto_ files as per commit.
Release Notes:
[Issue 6589](https://github.com/envoyproxy/envoy/issues/6589) allows for optional normalization of upstream requests independent of the normalization settings received from downstream.  Upstream normalization defaults to the _normalize_path_ setting and can be overridden with the setting _normalize_path_for_upstream_. 

It is imperative that existing functionality/behaviour is preserved. As an example,  the continued support for 0/100 in the runtime configuration for the normalize_path* properties in the http listener, while introducing support for Boolean true/false case.

Fixes #Issue 6589

There are several consequences to implementing support for this new property:

- As of proto3 defaults can not be specified in the `.proto` file.  This will require some changes to the initialisation code.  Will have to be sure to check all places where class is loaded. 
- Will need to code for the runtime case (see runtime.rst)  The `runtime.rst` in `root/configuration/http_conn_man` refers to % in relation to normalize_path.  Harvey's note in issue ~config.cc~ (see next task).  Support backward compatibility.
// TODO(htuch): we should have a boolean variant of featureEnabled() here.
- `ENVOY_NORMALIZE_PATH_BY_DEFAULT` `envoy/blob/master/source/extensions/filters/network/http_connection_manager/config.cc`
test for `ENVOY_NORMALIZE_PATH_BY_DEFAULT` line 205 should this be modified with
a boolean? (Needs a boolean featureEnabled() method as per HT's comment.)  When the adding the featureEnabled(absl::string_view, bool) ensure that specify no default conversion from int to boolean.  Include support for the `ENVOY_NORMALIZE_PATH_FOR_UPSTREAM_BY_DEFAULT`.
There is already support for `getBoolean(absl::string_view, bool default_value)`
in `envoy/source/common/runtime/runtime_impl.cc`. This can be reused. NOTE:
Support of existing functionality must be maintained (ie 0/100 -> false/true).
- `add normalize_path_for_upstream_` to the constructor `HttpConnectionManagerConfig::HttpConnectionManagerConfig()` in `envoy/blob/master/source/extensions/filters/network/http_connection_manager/config.cc`
- http manager `config.h` Need to add `normalize_path_for_upstream` to `envoy/source/extensions/filters/network/http_connection_manager/config.h`

### CODE REVIEW

File: api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto [hunk 1]
```diff
@@ -483,16 +483,32 @@ message HttpConnectionManager {
   // Should paths be normalized according to RFC 3986 before any processing of
   // requests by HTTP filters or routing? This affects the upstream *:path* header
   // as well. For paths that fail this check, Envoy will respond with 400 to
-  // paths that are malformed. This defaults to false currently but will default
-  // true in the future. When not specified, this value may be overridden by the
-  // runtime variable
+  // paths that are malformed. This defaults to true.
```
Comment (mattklein123) [inline api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:486]: v2 is frozen so please make the changes in v3 only.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 2]
```diff
@@ -478,16 +478,33 @@ message HttpConnectionManager {
...
+  // The upstream path normalization can be specified independently, see
+  // :ref:`normalize_path_for_upstream
+  // <envoy_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.normalize_path_for_upstream>`.
   // See `Normalization and Comparison <https://tools.ietf.org/html/rfc3986#section-6>`
   // for details of normalization.
   // Note that Envoy does not perform
   // `case normalization <https://tools.ietf.org/html/rfc3986#section-6.2.2.1>`
   google.protobuf.BoolValue normalize_path = 30;
 
+  // Should paths to upstream be normalized according to RFC 3986 after any processing of
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:494]: A few questions to get started:
1) When does this return 400 and if we want to return a 400 why would someone do this here vs. just doing it at the beginning before filters run? It's not clear to me what utility this has.
2) Should we actually deprecate normalize_path and have a new field that is an enum that describes which normalizing approach to use? Also, in the above docs it says that normalize_path defaults to true. I thought it defaults to false or maybe I am confused?

File: include/envoy/http/header_map.h [hunk 3]
```diff
@@ -310,6 +310,7 @@ class HeaderEntry {
   HEADER_FUNC(OtSpanContext)                                                                       \
   HEADER_FUNC(Origin)                                                                              \
   HEADER_FUNC(Path)                                                                                \
+  HEADER_FUNC(OriginalPathForUpstream)                                                             \
```
Comment (mattklein123) [inline include/envoy/http/header_map.h:313]: Do we need a new header? If the path is change can we still just use EnvoyOriginalPath above?

File: api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto (hunk 1 above)
Comment (twghu) [reply api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:486]: Ack.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 2 above)
Comment (twghu) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:494]: 1.  Agreed, yes that makes sense.
2.  The addition on a the extra header was intended to minimize the impact on routing/filtering code changes to deal with potential changes to, as you say, replacing normalize_path with an enum.  Idealy yes an enum is a cleaner approach.

File: include/envoy/http/header_map.h (hunk 3 above)
Comment (twghu) [reply include/envoy/http/header_map.h:313]: Ah, the EnvoyOriginalPath, I was curious about using that.  The issue as I understand it, is that ideally keeping everything "backward compatible" was a high priority.  So leaving normalize_path in situ as is, with minimal processing changes presented a lower risk option.  I'd prefer to re-use what code variables already exist rather than adding a new header.  If we can re-use the EnvoyOriginalPath for this then all the better.  I will dig into it's usage.  If I understand the requirement correctly, it would seem that once filtering and routing phase is complete then the could be set to the upstream normalized/non-normalized value depending on settings.  The caveat here is extra checking of the OriginalEnvoyPath might be needed for any code that relies on the original usage of the Path variable (will verify usage).

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:494]: A deprecation makes sense IMHO, as long as we preserve existing behavior today.

File: include/envoy/http/header_map.h (hunk 3 above)
Comment (htuch) [reply include/envoy/http/header_map.h:313]: One thing I worry about is potential misuse and inconsistent use by filters. Ideally we only make use of the original path when sending to upstream..

File: api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto [hunk 4]
```diff
@@ -483,16 +483,32 @@ message HttpConnectionManager {
...
+  // For paths that fail this check, Envoy will respond with 400 to
+  // paths that are malformed. This defaults to the value of
+  // :ref:`normalize_path
+  // <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.normalize_path>`.
+  // When not specified, this value may be overridden by the runtime variable
+  // :ref:`http_connection_manager.normalize_path_for_upstream<config_http_conn_man_runtime_normalize_path_for_upstream>`.
+  // See `Normalization and Comparison <https://tools.ietf.org/html/rfc3986#section-6>`
+  // for details of normalization.
+  // Note that Envoy does not perform
+  // `case normalization <https://tools.ietf.org/html/rfc3986#section-6.2.2.1>`
```
Comment (htuch) [inline api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto:509]: `[#not-implemented-hide:]`?

File: api/envoy/extensions/filters/network/http_connection_manager/v4alpha/http_connection_manager.proto [hunk 5]
```diff
@@ -478,16 +478,33 @@ message HttpConnectionManager {
   // Should paths be normalized according to RFC 3986 before any processing of
   // requests by HTTP filters or routing? This affects the upstream *:path* header
   // as well. For paths that fail this check, Envoy will respond with 400 to
-  // paths that are malformed. This defaults to false currently but will default
-  // true in the future. When not specified, this value may be overridden by the
-  // runtime variable
+  // paths that are malformed. This defaults to true.
+  // This by default affects the upstream *:path* header as well.
```
Comment (alyssawilk) [inline api/envoy/extensions/filters/network/http_connection_manager/v4alpha/http_connection_manager.proto:482]: Above already says This affects the upstream *:path* header

Maybe just change the first mention to include default?

Comment (alyssawilk) [inline api/envoy/extensions/filters/network/http_connection_manager/v4alpha/http_connection_manager.proto:482]: Actually reading this I was really confused about what was meant by affecting the upstream path header.  Can we maybe rephrase to contract normalizing path for in-Envoy code, vs normalizing the path *sent* uptream?

File: include/envoy/http/header_map.h (hunk 3 above)
Comment (mattklein123) [reply include/envoy/http/header_map.h:313]: Yeah I think we should use a single header here. Re: back compat, as this is a new feature, and controlled by an enum, IMO we can make it clear what that header holds in each enum case.
